### ISSUE TITLE
Should a transitive use ever allow access to a private symbol

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20707]
Should the below program compile? If it does, `public use M;` in `module Other` must be bringing in private symbols (and then whether the private symbols can be used or not is decided later). But it would seem more intuitive for that use statement not to bring in private symbols because `Other` does not have access to them.

``` chapel
module M {
  private var x: int;
  private proc f() {
    writeln("in f");
  }
  module Sub {
    use Other;
    proc inner() {
      writeln("x is ", x);
      f();
    }
  }
  proc main() {
    use Sub;
    inner();
  }
}

module Other {
  public use M;
}
```

**Comment by mppf** [comment]
@riftEmber - I made this issue about a case we were talking about

**Comment by mppf** [comment]
@lydia-duncan - any thoughts here?

**Comment by bradcray** [comment]
> Should the below program compile?

I don't think it should.  Or at least would need some sort of argument for why it should / what `private` should mean if it were to.

**Comment by lydia-duncan** [comment]
Yeah, I agree with Brad.  It's only valid for Sub to have access to the private symbols of M if it uses M directly (since it is in the same scope) or if it uses/imports a module that itself has access to those private symbols.  It seems like it would be wrong for a module's `public use` to enable access to symbols that it itself can't reference

**Comment by mppf** [comment]
I've marked this as Chapel 2.0 because the above program *does* compile today with the production compiler so changing it to an error is breaking. However I expect it is very unlikely that user code will have this pattern.

**Comment by bradcray** [comment]
We can still fix bugs after Chapel 2.0, even if they cause programs that had been sneaking through to change in behavior or stop working.  So I don't know that this needs to be considered a 2.0 issue as long as we've documented the intended behavior correctly (which I don't happen to know offhand).
